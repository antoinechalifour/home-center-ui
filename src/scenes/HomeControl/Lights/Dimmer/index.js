import { graphql, compose } from 'react-apollo'
import { updateLight, toggleLight } from 'queries/lights'
import Dimmer from './Dimmer'

const updateLightOptions = {
  name: 'updateLight'
}

const toggleLightOptions = {
  name: 'toggleLight',
  options: {
    refetchQueries: ['GetLights']
  }
}

export default compose(
  graphql(updateLight, updateLightOptions),
  graphql(toggleLight, toggleLightOptions)
)(Dimmer)
