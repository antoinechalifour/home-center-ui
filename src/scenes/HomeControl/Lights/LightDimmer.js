import React from 'react'

const LightDimmer = () => <div>LightDimmer</div>

export default LightDimmer
