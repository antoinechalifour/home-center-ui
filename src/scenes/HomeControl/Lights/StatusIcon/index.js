import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Icon from 'react-icons/lib/md/lightbulb-outline'

export default function StatusIcon ({ isOn, onClick }) {
  return (
    <IconWrapper ison={isOn}>
      <Icon onClick={onClick} />
    </IconWrapper>
  )
}

StatusIcon.propTypes = {
  isOn: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired
}

const IconWrapper = styled.span`
  svg {
    cursor: pointer;
    font-size: 50px;
    margin-right: 12px;

    opacity: ${({ theme, ison }) => (ison ? 1 : 0.24)};
    color: ${({ theme, ison }) => (ison ? theme.colors.primary : theme.colors.textInverse)};
  }
`
