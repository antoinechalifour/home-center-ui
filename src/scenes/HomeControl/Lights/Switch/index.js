import { graphql, compose } from 'react-apollo'
import { toggleLight, updateLight } from 'queries/lights'
import Switch from './Switch'

const toggleLightOptions = {
  name: 'toggleLight',
  options: {
    refetchQueries: ['GetLights']
  }
}

const updateLightOptions = {
  name: 'updateLight',
  options: {
    refetchQueries: ['GetLights']
  }
}

export default compose(
  graphql(toggleLight, toggleLightOptions),
  graphql(updateLight, updateLightOptions)
)(Switch)
