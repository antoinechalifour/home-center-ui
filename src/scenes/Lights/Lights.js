import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import gqlLoaderHoc from 'components/GqlLoader'
import Title from 'ui/WidgetTitle'
import SwitchableLight from './SwitchableLight'
import DimmableLight from './DimmableLight'

export function Lights ({ data }) {
  const renderers = {
    switch: props => <SwitchableLight {...props} />,
    dimmer: props => <DimmableLight {...props} />
  }

  return (
    <div>
      <Title>My devices</Title>
      <ul>
        {data.lights &&
          data.lights.map(x => {
            const renderLight = renderers[x.type]
            return (
              <Li key={x.id}>
                {renderLight(x)}
              </Li>
            )
          })}
      </ul>
    </div>
  )
}

Lights.propTypes = {
  data: PropTypes.shape({
    lights: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired,
        status: PropTypes.string,
        bri: PropTypes.number
      })
    )
  }).isRequired
}

export default gqlLoaderHoc(Lights)

const Li = styled.li`
  padding: 24px 12px;
  
  + li {
    border-top: 1px solid rgba(255, 255, 255, .15);
  }
`
