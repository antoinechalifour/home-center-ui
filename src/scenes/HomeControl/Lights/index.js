import { graphql } from 'react-apollo'
import { getLights } from 'queries/lights'
import Lights from './Lights'

export default graphql(getLights)(Lights)
