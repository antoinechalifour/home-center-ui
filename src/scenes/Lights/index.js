import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { graphql } from 'react-apollo'
import { getLights, lightUpdated } from 'queries/lights'
import Lights from './Lights'

export class LightsContainer extends Component {
  static propTypes = {
    data: PropTypes.shape({
      startPolling: PropTypes.func.isRequired,
      stopPolling: PropTypes.func.isRequired
    }).isRequired
  }

  componentDidMount () {
    this.props.data.subscribeToMore({
      document: lightUpdated,
      updateQuery: this._onLightUpdated
    })
  }

  _onLightUpdated = (prev, { subscriptionData }) => {
    if (!subscriptionData.data) {
      return
    }

    const updatedLight = subscriptionData.data.lightUpdated

    return {
      ...prev,
      lights: prev.lights.map(light => {
        if (light.id === updatedLight.id) {
          return updatedLight
        }

        return light
      })
    }
  }

  render () {
    return <Lights {...this.props} />
  }
}

export default graphql(getLights)(LightsContainer)
