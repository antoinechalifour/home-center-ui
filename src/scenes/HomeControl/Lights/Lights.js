import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Loader from 'ui/Loader'
import Switch from './Switch'
import Dimmer from './Dimmer'

export default function Lights ({ data }) {
  const renderers = {
    switch: props => <Switch {...props} />,
    dimmer: props => <Dimmer {...props} />
  }

  if (data.loading) {
    return <Loader />
  }

  return (
    <List>
      {data.lights &&
        data.lights.map(x => {
          const renderLight = renderers[x.type]
          return <Li key={x.id}>{renderLight(x)}</Li>
        })}
    </List>
  )
}

Lights.propTypes = {
  data: PropTypes.shape({
    loading: PropTypes.bool.isRequired,
    lights: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        status: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired
      })
    )
  }).isRequired
}

const List = styled.ul`
  padding: 24px;
`

const Li = styled.li`
  padding: 12px 0;
`
