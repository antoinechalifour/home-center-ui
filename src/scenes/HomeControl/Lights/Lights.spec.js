import 'jest-styled-components'
import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import Lights from './Lights'

const renderer = new ShallowRenderer()

describe('scenes/HomeControl/Lights', () => {
  it('Should render a loader when the data is not fetched', () => {
    const props = {
      data: {
        loading: true
      }
    }
    const tree = renderer.render(<Lights {...props} />)

    expect(tree).toMatchSnapshot()
  })

  it('Should render the list of lights otherwise', () => {
    const props = {
      data: {
        loading: false,
        lights: [
          { id: '1', type: 'switch', name: '#1', status: 'on' },
          { id: '2', type: 'dimmer', name: '#2', status: 'on' },
          { id: '3', type: 'switch', name: '#3', status: 'off' }
        ]
      }
    }
    const tree = renderer.render(<Lights {...props} />)

    expect(tree).toMatchSnapshot()
  })
})
