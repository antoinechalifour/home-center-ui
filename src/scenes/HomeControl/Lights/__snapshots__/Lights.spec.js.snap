// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scenes/HomeControl/Lights Should render a loader when the data is not fetched 1`] = `<Loader />`;

exports[`scenes/HomeControl/Lights Should render the list of lights otherwise 1`] = `
<styled.ul>
  <styled.li>
    <styled.div>
      <styled.div>
        <Apollo(Apollo(Switch))
          id="1"
          name="#1"
          status="on"
          type="switch"
        />
      </styled.div>
    </styled.div>
  </styled.li>
  <styled.li>
    <styled.div>
      <styled.div>
        <Apollo(Dimmer)
          id="2"
          name="#2"
          status="on"
          type="dimmer"
        />
      </styled.div>
    </styled.div>
  </styled.li>
  <styled.li>
    <styled.div>
      <styled.div>
        <Apollo(Apollo(Switch))
          id="3"
          name="#3"
          status="off"
          type="switch"
        />
      </styled.div>
    </styled.div>
  </styled.li>
</styled.ul>
`;
